feat(Ruby): TiC draft (publish with 8.6.0 eta 04-04-22) (#6479)
newrelic/docs-website: src/content/docs/logs/logs-everywhere/ruby-telemetry-context.mdx

  **Optional adjustments:**

  Once this is enabled, you also have control over the maximum number of logs sent to New Relic every minute. The default value is 10,000. If more than 10,000 logs are received in a 60-second window, your logs will begin to be sampled.
  
  Set it to a higher number to receive more logs. Set it to a lower number to receive fewer logs. Any integer is valid.

  newrelic.yml:
  ```
  application_logging.forwarding.max_samples_stored: 10_000
  ```

  Environment variable:
  ```
  NEW_RELIC_APPLICATION_LOGGING_FORWARDING_MAX_SAMPLES_STORED: 10000
  ```
  </Collapser>

  <Collapser
    id="2-decorate"
    title="Option 2: Let the agent decorate your logs."
  >

  Already have a log forwarder you like? We've got you covered. Language agents can decorate your logs with the linking metadata needed to provide access to telemetry-in-context features.
  
  Simply enable log decorating in your configuration, and relaunch the agent to start decorating your logs.
  
  newrelic.yml:
  ```
  application_logging.local_decorating.enabled: true
  ```

  Environment variable:
  ```
  NEW_RELIC_APPLICATION_LOGGING_LOCAL_DECORATING_ENABLED=true
  ```
  

>   Decoration is made possible by the addition of four attributes: `entity.guid`, `hostname`, `span.id`, and `trace.id`.

Reviewer: ```suggestion
  Our decorator adds four attributes to every log message: `entity.guid`, `hostname`, `trace.id`, and `span.id`.
  
  Example: 
  
  `This is my log message. NR-LINKING|{entity.guid}|{hostname}|{trace.id}|{span.id}|`
```

  ![Use infrastructure agent as log forwarder](./images/infra-log-forwarder.png "Use infrastructure agent as log forwarder")

  If you don't have any log forwarder enabled, the UI will prompt you to use the infrastructure agent as your log forwarder.
  </Collapser>
</CollapserGroup>

## Secure your data [#obfuscation]

Your logs may include sensitive information protected by HIPAA or other compliance protocols. By default we obfuscate number patterns that appear to be for items such as credit cards or Social Security numbers, but you may need to hash or mask additional information.

For more information, see our documentation about [obfuscation expressions and rules](https://docs.newrelic.com/docs/logs/ui-data/obfuscation-ui/). You can hash or mask your log data by using the New Relic UI or by using NerdGraph, our GraphQL API.

## Explore your data [#view-ui]

To make the most of your logging data:

* On the [APM **Summary** page](https://docs.newrelic.com/docs/apm/apm-ui-pages/monitoring/apm-summary-page-view-transaction-apdex-usage-data), click your **Web transaction time** chart to view logs associated with a specific point in time.

> * Check your app's [**Errors inbox**](https://docs.newrelic.com/docs/errors-inbox/errors-inbox) to view the logs associated with your errors.

Reviewer: "Errors Inbox" with a capital "I" is how it's written in the UI. Not sure if that's grammatically correct though.
```suggestion
* Check your app's [**Errors Inbox**](https://docs.newrelic.com/docs/errors-inbox/errors-inbox) to view the logs associated with your errors.
```

* Use [distributed tracing](https://docs.newrelic.com/docs/distributed-tracing/ui-data/understand-use-distributed-tracing-ui) to see the logs associated with individual traces.

> * Set up [alerts](/docs/alerts-applied-intelligence/new-relic-alerts/alert-conditions/create-alert-conditions/).

Reviewer: I think this could use some additional logging sparkle. Maybe something like this?

```suggestion
* Set up [alerts](/docs/alerts-applied-intelligence/new-relic-alerts/alert-conditions/create-alert-conditions/) based on log output and severity.
```

> * [Query your data](/docs/query-your-data/explore-query-data/get-started/introduction-querying-new-relic-data/) and [create dashboards](/docs/query-your-data/explore-query-data/dashboards/introduction-dashboards/).

Reviewer: Since the logs query page is different from the querying page used for metrics and events, I think it might be better to link directly to that document:

```suggestion
* [Query your data](/docs/logs/ui-data/query-syntax-logs/) using our specialized UI for logs data and [create dashboards](/docs/query-your-data/explore-query-data/dashboards/introduction-dashboards/) with the results.
```

* Explore more logging data across your platform with our [Logs UI](/docs/logs/log-management/ui-data/use-logs-ui/). For example, with telemetry in context, you can easily see [log patterns](/docs/logs/ui-data/find-unusual-logs-log-patterns), or show [surrounding logs](/docs/logs/troubleshooting/find-issues-cause-or-impact-surrounding-logs) at the time an error occurred in your app.

## Troubleshooting tips [#troubleshooting]

> Typically your logs will start to appear in less than a minute after you enable telemetry in context. If you don't see any logs for errors or traces, there may not be any for your app. Try refreshing the UI page, or change the [selected time period](https://docs.newrelic.com/docs/new-relic-one/ui-data/basic-ui-features/#data-analysis).

Reviewer: Love this phrasing!

* Cut through the noise of thousands of logs when troubleshooting time-critical issues, so you automatically see only the most relevant logs.
* Navigate within multiple types of telemetry data, and have the data correlate back to the original issue.
* Easily drill down into more detailed information from the same place in the UI.
* Find the log lines that you need to identify and resolve a problem.

![Telemetry in context in APM UI](./images/ruby-errors-log-details.png "Telemetry in context in APM UI")

<figcaption>
  **Telemetry in context:** From the left nav in the APM UI, you can select an error and immediately see related logs, log details, host name, important attributes, and more.
</figcaption>

## Enable and configure [#enable]

You have three options to configure telemetry in context for your application:

<CollapserGroup>

  <Collapser
    id="1-agent"
    title="Option 1 (simplest): Let the agent forward your logs."
  >

  This is the simplest approach, and it's a great choice for developers who may not have the access or interest in setting up a log forwarder, or for accounts that want to see the power of logs in context without a lot of overhead. 

>   All you need to do is install an agent version with log forwarding capabilities ([Ruby agent 6.8.0 or higher](https://docs.newrelic.com/docs/release-notes/agent-release-notes/ruby-release-notes/)), and update your configuration to enable forwarding.

Reviewer: Whoops, I think the numbers got swapped!

```suggestion
  All you need to do is install an agent version with log forwarding capabilities ([Ruby agent 8.6.0 or higher](https://docs.newrelic.com/docs/release-notes/agent-release-notes/ruby-release-notes/)), and update your configuration to enable forwarding.
```

  newrelic.yml:
  ```

>   application_logging.forwarding.enabled: true

Reviewer 2: Do we typically show config keys for YAML with dotted notation vs the nested form?

```
application_logging:
  forwarding:
    enabled: true
```

Reviewer: Great point, @jasonrclark! We typically do not. I don't know why I chose to write it in that form that time around. I think it's better to use the nested form.

* Find the log lines that you need to identify and resolve a problem.

![Telemetry in context in APM UI](./images/ruby-errors-log-details.png "Telemetry in context in APM UI")

<figcaption>
  **Telemetry in context:** From the left nav in the APM UI, you can select an error and immediately see related logs, log details, host name, important attributes, and more.
</figcaption>

## Enable and configure [#enable]

You have three options to configure telemetry in context for your application.

<CollapserGroup>

  <Collapser
    id="1-agent"
    title="Option 1 (simplest): Let the agent decorate and forward your logs."
  >

  This is the simplest approach, and it's a great choice for developers who may not have the access or interest in setting up a log forwarder, or for accounts that want to see the power of logs and other telemetry in context without a lot of overhead.
  
  Using this option, your logs will include `span.id`, `trace.id`, `hostname`, `entity.guid`, and `entity.name`. These attributes will automatically link your logs with spans, traces, and errors for your app, its associated hosts, and other entities.

>   All you need to do is install an agent version with log forwarding capabilities ([Ruby agent 8.6.0 or higher](https://docs.newrelic.com/docs/release-notes/agent-release-notes/ruby-release-notes/)), and update your configuration to enable forwarding.

Reviewer: @tannalynn  - this is where you update the release notes link for 8.6.0

Reviewer 3: Now that the release notes PR has been created. The link can be updated to 
https://docs.newrelic.com/docs/release-notes/agent-release-notes/ruby-release-notes/ruby-agent-850/

---
title: 'Ruby telemetry in context'
metaDescription: For apps monitored by the Ruby agent, learn how to link log data with related data across the rest of the New Relic platform.
---

> **REVIEWER: My biggest concern is that I'm throwing way too many words into this doc. I can't decide if it's a good idea to focus on saying why this matters, or does it go too far? Would a better approach be to emphasize how simple it is? For example, if it really is just 1-update to the latest agent, and 2-do any optional config you want to do, maybe we can have one doc for all the agents? Please feel free to comment or edit directly in my PR. Thanks!**

Reviewer: I like including why this matters/how to use it. If something is going to be enabled for me, I want simple direction on how to use it. This is where UI screenshots may also come in handy! 

I think this could work well as one doc for all the agents, as long as there were either per-language links or a dropdown menu to break down the configuration. I imagine users would find it convenient to have the content on this page (though the release notes route could also work!)

Author: We can experiment both ways. My goal is to have the fewest number of docs to make it look as easy as possible!

* Explore the logging data across your platform with our [Logs UI](/docs/logs/log-management/ui-data/use-logs-ui/).
* See your logs in context of your app's performance in the [APM UI](/docs/logs/log-management/ui-data/use-logs-ui/#links). Troubleshoot [errors](/docs/apm/apm-ui-pages/error-analytics/errors-page-find-fix-verify-problems/) with [distributed tracing](/docs/apm/distributed-tracing/ui-data/understand-use-distributed-tracing-data), stack traces, application logs, and more.
* Get deeper visibility into both your application and your platform performance data by forwarding your logs with our [infrastructure monitoring agent](/docs/logs/enable-log-management-new-relic/enable-log-monitoring-new-relic/forward-your-logs-using-infrastructure-agent/). Review your [infrastructure logs](/docs/logs/log-management/ui-data/use-logs-ui/#links) in the UI.
* Set up [alerts](/docs/alerts-applied-intelligence/new-relic-alerts/alert-conditions/create-alert-conditions/).
* [Query your data](/docs/query-your-data/explore-query-data/get-started/introduction-querying-new-relic-data/) and [create dashboards](/docs/query-your-data/explore-query-data/dashboards/introduction-dashboards/).

> **REVIEWER: These bullets come from the original Ruby logs in context doc. Will need to be updated.**

Reviewer 2: Seems like we're going to want to point to some of the newer in-context APM/Errors/etc. UI pieces where we're adding logging information vs just always pointing to the generic Log Explorer docs. Not sure the right spot for all of that at this point.

Author: Absolutely right! I have some log info already in some APM docs. Am planning to update them with the new workflow. I'll rework this section to make the current experience more clear. Thanks for the reminder!

> ## Optional configuration [#config]

Reviewer: I would recommend moving this below the `install or update to...` it's hard to skim for here. I think an engineer would probably want to read about installation and configuration in the same area. 

I also think there could be some benefit from having config information in this section or linking to the future config entries on Ruby's config page.

Author: Yes, thanks! Am moving this with another commit to branch

Author: Also, please keep me in the loop about the actual config values so I can cross-link to the Ruby config doc when ready.

> When you update to telemetry in context, three changes are made to your configuration file:

Reviewer 2: Not following if this is trying to say "these are things you can configure" or "things that are defaulted now for you"

Author: I didn't word this well. Let's chat about how to make this more clear!

* Application log forwarding enabled by default

> * Maximum log sample stored (10,000 lines by default)

Reviewer: 10,000 lines per minute

Author: Great catch, thanks!

> * Logging metrics, including `trace.id`, `span.id`, `entity_guid`, and `hostname`

Reviewer 2: This isn't really about configuration -- this is the effect that we have by doing telemetry-in-context. The attributes get applied to log lines, though, not the metrics (but not sure this belongs here).

```suggestion
* Logging metrics, including `trace.id`, `span.id`, `entity.guid`, and `hostname`
```

Author: Clearly I got things mixed up. Will need your help to reword!

To collect other metadata, or to change these default settings, update your [`newrelic.yml` configuration file](https://docs.newrelic.com/docs/apm/agents/ruby-agent/configuration/ruby-agent-configuration/).

> **REVIEWER: Does anything else in the [original Ruby logs in context doc](https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-ruby) need to be added to this?**

Reviewer: I'd like to take a closer look at the [incompatible gems](https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-ruby/#incompabilities) section to see if they are also incompatible with this feature. Some of them I haven't heard about/tested before. There's another one I recently learned about called Log4r, which is like Log4j but for Ruby.  

As an aside, do you know if there's an intention to remove support for the DecoratingLogger? I wonder if anyone might still need access to that information if they're using that procedure? (Perhaps @jasonrclark would know?)

Author: I have not heard anything specific about the other ones listed in https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-ruby . Do you know, Jason? Originally I was thinking that we could say something along the lines of, "If necessary, you can use the original logs in context options [link to this doc] but be aware that you will not have as much telemetry data with those options." or something like that.

---

[Telemetry in context](https://docswebsitedevelop-doc7761logsintro.gtsb.io/docs/logs/logs-everywhere/telemetry-in-context/) for the Ruby agent connects your logs with all of your telemetry data for your apps, hosts, and other entities. Bringing all of this data together in a single tool helps you quickly:

* Cut through the noise of thousands of logs when troubleshooting time-critical issues, so you automatically see only the most relevant logs.
* Navigate within multiple types of telemetry data, and have the data correlate back to the original issue.
* Easily drill down into more detailed information from the same place in the UI.
* Find the log lines that you need to identify and resolve a problem.

![Telemetry in context in APM UI](./images/ruby-errors-log-details.png "Telemetry in context in APM UI")

<figcaption>
  **Telemetry in context:** From the left nav in the APM UI, you can select an error and immediately see related logs, log details, host name, important attributes, and more.
</figcaption>

## Install and configure [#enable]

> To enable telemetry in context for APM apps monitored by Ruby, all you need to do is:
> 
> 1. Install or update to [Ruby agent version 8.6.0 or higher](https://docs.newrelic.com/docs/release-notes/agent-release-notes/ruby-release-notes). **REVIEWER: need to confirm version**
> 2. To collect other metadata, or to change default settings, update your [`newrelic.yml` configuration file](https://docs.newrelic.com/docs/apm/agents/ruby-agent/configuration/ruby-agent-configuration/). **REVIEWER: what optional configuration might they want to do?**
> 
> When you update to telemetry in context, two changes are made to your configuration file:
> 
> * Application log forwarding enabled by default
> * Maximum log sample stored (approximately 10,000 lines per minute by default)
> 
> In addition, log decoration by default now includes `trace.id`, `span.id`, `entity.guid`, and `hostname`.

Reviewer: ```suggestion
* **Option 1: Let the agent forward your logs**

This is the simplest approach. All you need to do is install a version of a language agent with log forwarding abilities, and update your configuration to enable forwarding. 

newrelic.yml: `application_logging.forwarding.enabled: true`
Environment variable: `NEW_RELIC_APPLICATION_LOGGING_FORWARDING_ENABLED=true`

This is a great choice for developers who may not have the access or interest in setting up a log forwarder or for accounts that want to see the power of logs in context without a lot of overhead.

This option may be on by default in a future version of the agent.

**Optional adjustments** 
Once this is enabled, you also have control over the maximum number of logs sent to New Relic every minute. The default value is 10,000. If more than 10,000 logs are received in a 60-second window, your logs will begin to be sampled. Set it to a higher number to receive more logs. Set it to a lower number to receive fewer logs. Any Integer is valid.

newrelic.yml: `application_logging.forwarding.max_samples_stored: 10_000`
Environment variable: `NEW_RELIC_APPLICATION_LOGGING_FORWARDING_MAX_SAMPLES_STORED: 10000`

* **Option 2: Let the agent decorate your logs**

Already have a log forwarder you like? We've got you covered. Language agents can now decorate your logs with the linking metadata needed to provide access to logs-in-context features. 

Simply enable log decorating in your configuration and relaunch the agent to start decorating your logs.

newrelic.yml: `application_logging.local_decorating.enabled: true`
Environment variable: `NEW_RELIC_APPLICATION_LOGGING_LOCAL_DECORATING_ENABLED=true`

Decoration is made possible by the addition of four attributes: `entity.guid`, `hostname`, `span.id, and `trace.id`. 

This option is preferred over the old decorating formatter, `NewRelic::Agent::Logging::DecoratingFormatter`.
 
* **Option 3: Use the old decorating formatter**  

Before language agents had the ability to forward and decorate logs, logs-in-context could be enabled by updating your application to use the `NewRelic::Agent::Logging::DecoratingFormatter` to send linking metadata. 

This option is still supported, but is no longer encouraged. For instructions to use this approach, visit [Ruby: Configure logs in context](https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-ruby/). 
```

> **REVIEWER: Anything to add about what to do if they don't see telemetry data? The original Ruby logs in context doc has a troubleshooting section. https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-ruby#troubleshooting**

Reviewer: ```suggestion
## Troubleshooting [#troubleshooting]

* How can I tell if it's working? 
The best place to check to see if log data associated with your application is the APM Logs left navigation option. Data should start flowing through five minutes after deploying any of the above configuration changes.

**BARB: (Screenshot with arrow pointing to the option?)**
```

## Explore your data [#view-ui]

To make the most of your logging data:

**REVIEWER: These bullets come from the original Ruby logs in context doc. Will need to be updated with more info about what they now see from APM and infrastructure.**

* Set up [alerts](/docs/alerts-applied-intelligence/new-relic-alerts/alert-conditions/create-alert-conditions/).
* [Query your data](/docs/query-your-data/explore-query-data/get-started/introduction-querying-new-relic-data/) and [create dashboards](/docs/query-your-data/explore-query-data/dashboards/introduction-dashboards/).

> * Explore more logging data across your platform with our [Logs UI](/docs/logs/log-management/ui-data/use-logs-ui/). For example, with telemetry in context, you can easily see [log patterns](/docs/logs/ui-data/find-unusual-logs-log-patterns), or show [surrounding logs](/docs/logs/troubleshooting/find-issues-cause-or-impact-surrounding-logs) at the time an error occurred in your app.

Reviewer: Other bullet ideas: 
* Use Errors Inbox to view the logs associated with your errors
* Visit Distributed Tracing to see the logs associated with individual traces
* Click on your Web Transaction Times chart on the APM Summary page to view logs associated with a specific point in time

(You likely know the best resources to link these pages to!)

> **REVIEWER: Does anything else in the [original Ruby logs in context doc](https://docs.newrelic.com/docs/logs/logs-context/configure-logs-context-ruby) need to be added to this? For example, saying something along the lines that those other options are still available if you cannot update to the current agent, but you won't enjoy all the rich telemetry data. Let's chat about how to word this, because really, the goal is to encourage customers to use the latest version.**

Reviewer: I linked to the original doc in **Option 3** above. Feel free to finesse that with more encouragement to use the agent as its forwarder!
